include/sound/sof.h

```diff
@@ -68,6 +68,9 @@ struct sof_dev_desc {
 
 	bool use_acpi_target_states;
 
+	/*set to 1 to not use ignore_machine and still use the original platform driver*/
+	bool no_ignore_machine;
```

Comment:
@yaochunhung I totally missed this. ignore_machine has nothing to do with ignoring the machine driver selected in SOF. It is more to do with how the core deals with the FE DAI links in the legacy machine driver. In your case, setting this should have no effect at all because there's no legacy driver to deal with right? @plbossart is my understanding correct?

Reply:
@ranj063 We still use legacy driver because some pcm nodes are still controlled by non-sof driver, not move to DSP. And We also want to keep to legacy BE DAI driver because the power and clock of our audio hardware are controlled by AP side. e.g ETDM2_OUT_BE. If no_ignore_machine is set as true, the legacy driver can be keep and reuse. Thanks!

Reply:
I plot the block diagram to explain our driver. Hope it more clear. Thanks!
  ------------------------------------------------------------------------------------------                                      
  | Normal ALSA pcm platform                                                                                      |                                      
  |  ---------------                                                                                ----------------    |                                      
  |  | DLx             |                                                                                | I2S playback |    |                                      
  |  |(normal FE)  | --------------------------------------------------->  | (normal BE)   |   |                                      
  |  ---------------                                                                                ----------------     |                                      
  |                                                            --------------------          -------------------     |                                               
  |  						               | SOF_DMA_DL2 |           | ETDM2_OUT_BE |    |         							             	                     
  |                                                            |   (widget)          |-----> | (normal BE)        |    |                                      
  |                                                            -- ----------------            --------------------    |                                      
  |                                                                            ^                                                      |                                      
  |                                                                             |                                                       |                                      
  ------------------------------------------------------------------------------------------                                      
                                                                                | Connect SOF DL2 to target SOF BE CPU DAI Widget’s source widget     
  --------------------------------------------------------------------------------------------                                     
  |  SOF ALSA pcm platform                                    |                                                       |                                     
  |                                                                             |                                                       |                                     
  |  -------------           ----------------                  --------------         ------------------    |                                     
  |  |  SOF_DL2  |----->| PCMxP          |---> ...--->| BUFn           |--->| AFE_SOF_DL2  |     |                                       
  |  | (SOF FE)    |          |(Host widget)|                  |(buf widget)|        | (SOF BE)         |     |                                     
  |  --------------          ----------------                  ---------------       ------------------     |                                     
  |                                                                                                                                      |                                     
  ------------------------------------------------------------------------------------------- --

Reply:
@yaochunhung can you draw this into something more user friendly? :)

Reply:
@dbaluta Sorry that I don't know how to attach file or diagram to github website. May I send the diagram to you, @plbossart, and @ranj063  dieclty by e-mail? Thanks!

Reply:
use e.g. https://asciiflow.com/#/ and copy-paste the diagram with backquotes to protect it.

Reply:
@plbossart  Thanks but I still have problem to copy-paste the diagram with backquotes. I share the diagram in the link. Thanks. https://asciiflow.com/#/share/eJzlWN2K2kAUfpVDrhQWL4Sl1Lv4xy5oV9BlbwIymtkaGichjlVZFvoQZaGv0cvSp8mT7Ex%2B3JlkEpOo3dYOQ0jGk%2FN952fOmfikEbTEWousbftKs9EOe1pLezK0raG1Pl5fXxnajt01P%2FA7ireUPRia%2F%2F33RU7DIOwKAPpgrMPYWRMTOsgzocJ4U%2FfuZp1wymZV8UvKSymVnxxviewwBu58Ca6N6CNbO0LlGVieXmWVcBzCKB1cBSmuqjvYiszZ2FjmZ0zBc9YUg4voQoS9bY5hxKrJDM2%2F7G1Vqq5F0e736pU8UHS%2B%2FGLXGKzNwXJo%2FR1hOJE%2FzpC1kKntrHj62rQc6HsOodAj5tnxaiHgDesAG%2BThugJvsrBW8VbYWLYNMwzINLEJFimNpxaQ8JZovrAIBtOzvmLvCohDAT0%2B4jmNpAtn7FFZWTU1U%2FaQcENKZT60rRG5Z3zXn3aH%2BrQ7aO7Re5PusDm9u59M2z1QcotypavfAuKZosJWO%2FzlpxgcORR%2BXEjUUZQZdLGLGfTNgyRUgEG%2BB2thtsUFLLhIdc3PPHKENbDDto%2Fn2DbL0dmuSm%2F7s1kmHQeYgeX5piOrMun0uV2MVY6ozEotcKFnzCI2A8tkwiufUCKAOkCRx8sxW2X5QjujeybMC8FDsHH8bz9WsHLW3hzHhVvAK81V5HOKmEOCTVTHhnpQSBKI7x6qM0XdT5aptN1Migc4qAqjzjDzIyHP12nUQCfLlboUibPu6pSoCrVkQh4SiWCOUyPrS1tVLvoKi8N9zfa0uClYrLcj0W3spn3fJ43tm5De703jdzP4xb4We4KCgRwWPsP2X7txVlRcbDT2P9baa3Yg8zLPCwcYJR6SjGo8Q%2Fs9qf1Hla0u%2BkRa46zCcljPYnD8%2FEfzFTLdrYpIuXEQ5GJmjpkZuV3Fl0qAi5wKU%2FOcccS4bD%2BmvOmfOi%2BTr%2B8BgqPEeJSQKMi1sGHlPJBB1bVcbPO%2FFXL0BX0sanvqxgdR4xN%2FDgD54buMVVkz%2B5zNljvO0nUIJlT0IexXhW6Zlsw%2FxL97yMp%2BX%2BQP%2Fz%2FY8Ib2rD2%2FAhvlfrM%3D)

Reply:
![image](https://user-images.githubusercontent.com/62316/132476344-923dfe3a-5305-43e5-9fc8-c63d9ab2c58f.png)

Reply:
Sorry, I have absolutely zero understanding of the suggested solution @yaochunhung 

The most puzzling part of the 'ETDM2_OUT_BE' which isn't going to be connected to the 'SOF_DL2 FE'. So when you start the 'SOF DL2 FE', how would e.g. the HW_PARAMS be propagated to the ETDM2_OUT_BE which controls power and clock?

Reply:
@plbossart  In the machine driver sound/soc/mediatek/mt8195/mt8195-mt6359-rt1019-rt5682.c. we define g_sof_conn_streams about the connection among ETDM2_OUT_BE,SOF_DMA_DL2 and AFE_SOF_DL2. 
static const struct sof_conn_stream g_sof_conn_streams[] = {
	{ "ETDM2_OUT_BE", "AFE_SOF_DL2", SOF_DMA_DL2, SNDRV_PCM_STREAM_PLAYBACK},
	{ "ETDM1_OUT_BE", "AFE_SOF_DL3", SOF_DMA_DL3, SNDRV_PCM_STREAM_PLAYBACK},
	{ "UL_SRC1_BE", "AFE_SOF_UL4", SOF_DMA_UL4, SNDRV_PCM_STREAM_CAPTURE},
	{ "ETDM2_IN_BE", "AFE_SOF_UL5", SOF_DMA_UL5, SNDRV_PCM_STREAM_CAPTURE},
};
Then we implement .late_probe = mt8195_mt6359_rt1019_rt5682_card_late_probe. 
In this function, we search pcm_runtime by sof_link and normal_link then get sof's pcm runtime : sof_rtd and normal pcm runtime : normal_rtd. After that, we search the sof's runtime cpu_dai then check the widget in its path then set dapm route's source and sink then add routes.
for example, ETDM2_OUT_BE is normal_link, AFE_SOF_DL2 is sof_link,SOF_DMA_DL2 is sof_dma, and stream_dir is SNDRV_PCM_STREAM_PLAYBACK. 
We will use snd_soc_dapm_widget_for_each_source_path to get the playback_widget is BUFn.x in the diagram above.
then set sink of dpam route as conn->sof_dma(SOF_DMA_DL2 in this case). After that, BUFn.x will be connected to SOF_DMA_DL2.
For SOF_DMA_DL2 to ETDM2_OUT_BE , we define the dapm route path in machine driver.
	/* SOF Downlink */
	{"I070", NULL, SOF_DMA_DL2},
	{"I071", NULL, SOF_DMA_DL2},
Then also define dapm route path in mt8195-dai-etdm.c
	{"O048", "I070 Switch", "I070"},
	{"O049", "I071 Switch", "I071"},
	{"ETDM2 Playback", NULL, "O048"},
	{"ETDM2 Playback", NULL, "O049"},
I hope this explanation is more clear. Thanks.

Reply:
> I hope this explanation is more clear. Thanks.

Humm, no :-(
I am missing the big picture, you just hammered us with details that are probably correct but don't really provide a sense of what the problem statement is, if you considered alternatives and what led you to pick this solution you're asking us to review.

Reply:
@plbossart  
We set no_ignore_machine as true to prevent ignoring normal FE in soc_check_tplg_fes.
 
/* ignore this FE */
if (dai_link->dynamic) {
dai_link->ignore = true;
continue;
})
 
The reason is that we still need to use normal FE for the paths which don't gothrough DSP. The benifit is that we can offload the process of certain audio data(e.g. speaker/headphone) to dsp and we can reuse normal alsa driver for those audio data which don't need dsp offloading.(e.g. Display port). Thanks.

Reply:
@yaochunhung It makes a bit more sense now. but a naive question. Is there a way for you define the Normal FW-> Normal BE in topology?

Reply:
@ranj063  For normal FE/BE, I think it need to communicate with DSP if defining them in topology(please correct me if my misunderstanding).  It takes lower effort to reuse normal FE/BE without topology. Thanks.